docs/design/sql/20-matching-sql-query-to-jet-job.md

```diff
@@ -0,0 +1,237 @@
+
+```java
+public interface SqlService {
+// ...
+
+    /**
+     * Execute the statement in asynchronous way if possible.
+     * @param statement
+     * @return
+     */
+    // also additional convenience variants similar to `execute`
+    // supported only on client-side (SqlClientService).
+    @Nonnull
+    SqlResult executeAsync(@Nonnull SqlStatement statement);
+}
+
+public interface AsyncSqlResult extends SqlResult {
+    /**
+     * @return If the query uses/used Jet job
+     */
+    boolean isJob();
+
+    /**
+     * Returns job id of job used to execute this query.
+     * <p>
+     * If the job was a light job (interactive query without {@code ANALYZE})
+     * it will be impossible to find the job after the query finished
+     * which requires caution or error handling to avoid race conditions.
+     *
+     * @throws UnsupportedOperationException if the query did not a use Jet job (isJob() returns false)
+     */
+    long getJobId();
+
+    /**
+     * Wait for query finish.
+     * If the job does not exist it assumes that it already finished.
+     *
+     * @return Final result of the query. It might be `this` or different object containing eg. final `updateCount` 
+     * @throws UnsupportedOperationException if the query did not a use Jet job (isJob() returns false)
+     */
+    SqlResult join();
+}
+
+// Only client-side API implements async execution mode
+public class SqlClientResult implements AsyncSqlResult {
+// ...
+}
+```
+
+Within the SqlService, we're implementing a new method named `executeAsync`. When using this method, the client can
+indicate that a query should be executed asynchronously by setting an asynchronous flag. Afterward, `execute` task
+is sent via calling `executeAsync`, the client will receive an initial `AsyncSqlResult` as part of the
+`SqlExecuteResponse`. This result will include the `jobId` of the corresponding Jet job. If the query produces rows,
+it will proceed with the regular execution process. However, if the user wishes to wait for query completion,
+they must explicitly call`AsyncSqlResult.join()`.
+
+The proposed async execution changes a response on DML query: after this change, the response will contain an
+empty cursor, which is tracked by `QueryClientState`.
+
+Also, it's important to note that asynchronous execution won't be the default mode, at least for the time being. This is
+because, without a flexible asynchronous mode, it could potentially introduce additional latency for short DML queries,
+as it involves an extra round-trip to the member.
+
+##### QueryClientStateRegistry
+
+Introduce `registerDml` for DML queries tracking by `QueryClientState`.
+
+##### PlanExecutor
+
+We should to introduce the same job workflow for DML queries for `PlanExecutor` as for `SELECT` queries : set up a
+job completion future in `QueryClientStateRegistry`.
+
+##### SqlExecuteMessageTask
+
+Encode response message with `jobId` if the query uses Jet job.
```

Comment:
this is obvious after documentation of client protocol changes and it is expected place for that - IMO not worth describing